File: java/src/main/java/io/cucumber/gherkin/GherkinTokenMatcher.java
package io.cucumber.gherkin;

import io.cucumber.gherkin.Parser.TokenMatcher;
import io.cucumber.messages.types.Location;
import io.cucumber.messages.types.StepKeywordType;
import org.jspecify.annotations.Nullable;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static io.cucumber.gherkin.Constants.COMMENT_PREFIX_CHAR;
import static io.cucumber.gherkin.Constants.DEFAULT_LANGUAGE;
import static io.cucumber.gherkin.Constants.DOCSTRING_ALTERNATIVE_SEPARATOR;
import static io.cucumber.gherkin.Constants.DOCSTRING_SEPARATOR;
import static io.cucumber.gherkin.Constants.TABLE_CELL_SEPARATOR;
import static io.cucumber.gherkin.Constants.TAG_PREFIX_CHAR;
import static io.cucumber.gherkin.Locations.COLUMN_OFFSET;
import static io.cucumber.gherkin.Locations.atColumn;
import static io.cucumber.gherkin.Parser.TokenType;

final class GherkinTokenMatcher implements TokenMatcher {

    private static final Pattern LANGUAGE_PATTERN = Pattern.compile("^#\\s*language\\s*:\\s*([a-zA-Z\\-_]+)\\s*$");
    private final String defaultLanguage;
    // Expect at most two languages, the default language and one other
    private final Map<String, KeywordMatcher> activeKeywordMatchers = new HashMap<>(2);
    private @Nullable String currentLanguage;
    private KeywordMatcher currentKeywordMatcher;
    private @Nullable String activeDocStringSeparator = null;
    private int indentToRemove = 0;

    GherkinTokenMatcher(String defaultLanguage) {
        this.defaultLanguage = defaultLanguage;
        this.currentKeywordMatcher = requireKeywordMatcher(defaultLanguage, new Location(0, 0));
        reset();
    }

    GherkinTokenMatcher() {
        this(DEFAULT_LANGUAGE);
    }

    @Override
    public void reset() {
        // TODO performance: reset() is called once in the constructor and once for each file (Parser.parse()). 
        // It could be called only once, but there is no measurable impact with the profiler
        activeDocStringSeparator = null;
        indentToRemove = 0;
        setLanguageMatched(defaultLanguage, null);
    }

    private void setLanguageMatched(String language, @Nullable Location location) {
        if (language.equals(currentLanguage)) {
            return;
        }
        KeywordMatcher keywordMatcher = requireKeywordMatcher(language, location);
        currentLanguage = language;
        currentKeywordMatcher = keywordMatcher;
    }

    private KeywordMatcher requireKeywordMatcher(String language, @Nullable Location location) {
        KeywordMatcher keywordMatcher = activeKeywordMatchers.computeIfAbsent(language, KeywordMatchers::of);
        if (keywordMatcher == null) {
            throw new ParserException.NoSuchLanguageException(language, location);
        }
        return keywordMatcher;
    }

    private void setTokenMatched(Token token, TokenType matchedType, @Nullable String text, @Nullable String keyword, int indent, @Nullable StepKeywordType keywordType, @Nullable List<LineSpan> items) {
        token.matchedType = matchedType;
        token.matchedKeyword = keyword;
        token.keywordType = keywordType;
        token.matchedText = text;
        token.matchedItems = items;
        token.matchedLanguage = currentLanguage;
        token.matchedIndent = indent;
        token.location = atColumn(token.location, token.matchedIndent + COLUMN_OFFSET);
    }

    @Override
    public boolean match_EOF(Token token) {
        if (!token.isEOF()) {
            return false;
        }
        setTokenMatched(token, TokenType.EOF, null, null, 0, null, null);
        return true;
    }

    @Override
    public boolean match_Other(Token token) {
        String text = removeDocStringIndent(token);
        setTokenMatched(token, TokenType.Other, unescapeDocString(text), null, 0, null, null);
        return true;
    }

    @Override
    public boolean match_Empty(Token token) {
        if (!token.getRequiredLine().isEmpty()) {
            return false;
        }
        setTokenMatched(token, TokenType.Empty, null, null, 0, null, null);
        return true;
    }

    @Override
    public boolean match_Comment(Token token) {
        Line line = token.getRequiredLine();
        if (!line.startsWith(COMMENT_PREFIX_CHAR)) {
            return false;
        }
        String text = line.getRawText();
        setTokenMatched(token, TokenType.Comment, text, null, 0, null, null);
        return true;
    }

    @Override
    public boolean match_Language(Token token) {
        Line line = token.getRequiredLine();
        if (!line.startsWith(COMMENT_PREFIX_CHAR)) {
            return false;
        }
        Matcher matcher = LANGUAGE_PATTERN.matcher(line.getText());
        if (!matcher.matches()) {
            return false;
        }
        String language = matcher.group(1);
        setTokenMatched(token, TokenType.Language, language, null, line.getIndent(), null, null);
        setLanguageMatched(language, token.location);
        return true;
    }

    @Override
    public boolean match_TagLine(Token token) {
        Line line = token.getRequiredLine();
        if (!line.startsWith(TAG_PREFIX_CHAR)) {
            return false;
        }
        List<LineSpan> tags = TagLine.parse(line.getIndent(), line.getText(), token.location);
        setTokenMatched(token, TokenType.TagLine, null, null, line.getIndent(), null, tags);
        return true;
    }

    @Override
    public boolean match_FeatureLine(Token token) {
        return matchTitleLine(token, TokenType.FeatureLine, currentKeywordMatcher::matchFeatureKeyword);
    }

    @Override
    public boolean match_RuleLine(Token token) {
        return matchTitleLine(token, TokenType.RuleLine, currentKeywordMatcher::matchRuleKeyword);
    }

    @Override
    public boolean match_BackgroundLine(Token token) {
        return matchTitleLine(token, TokenType.BackgroundLine, currentKeywordMatcher::matchBackgroundKeyword);
    }

    @Override
    public boolean match_ScenarioLine(Token token) {
        return matchTitleLine(token, TokenType.ScenarioLine, currentKeywordMatcher::matchScenarioKeyword);
    }

    @Override
    public boolean match_ExamplesLine(Token token) {
        return matchTitleLine(token, TokenType.ExamplesLine, currentKeywordMatcher::matchExampleKeyword);
    }

    private boolean matchTitleLine(Token token, TokenType tokenType, Function<Line, KeywordMatcher.@Nullable Match> matcher) {
        Line line = token.getRequiredLine();
        KeywordMatcher.Match match = matcher.apply(line);
        if (match == null) {
            return false;
        }
        String keyword = match.getKeyword();
        int keywordLength = match.getKeywordLength();
        String title = line.substringTrimmed(keywordLength);
        setTokenMatched(token, tokenType, title, keyword, line.getIndent(), null, null);
        return true;
    }

    @Override
    public boolean match_DocStringSeparator(Token token) {
        return activeDocStringSeparator == null
                // open
                ? match_DocStringSeparator(token, DOCSTRING_SEPARATOR, true) ||
                match_DocStringSeparator(token, DOCSTRING_ALTERNATIVE_SEPARATOR, true)
                // close
                : match_DocStringSeparator(token, activeDocStringSeparator, false);
    }

    private boolean match_DocStringSeparator(Token token, String separator, boolean isOpen) {
        Line line = token.getRequiredLine();
        if (!line.startsWith(separator)) {
            return false;
        }
        String mediaType = null;
        if (isOpen) {
            mediaType = line.substringTrimmed(separator.length());
            activeDocStringSeparator = separator;
            indentToRemove = line.getIndent();
        } else {
            activeDocStringSeparator = null;
            indentToRemove = 0;
        }

        setTokenMatched(token, TokenType.DocStringSeparator, mediaType, separator, line.getIndent(), null, null);
        return true;
    }

    @Override
    public boolean match_StepLine(Token token) {
        Line line = token.getRequiredLine();
        KeywordMatcher.StepMatch match = currentKeywordMatcher.matchStepKeyword(line);
        if (match == null) {
            return false;
        }
        String keyword = match.getKeyword();
        int keywordLength = match.getKeywordLength();
        StepKeywordType keywordType = match.getKeywordType();
        String stepText = line.substringTrimmed(keywordLength);
        setTokenMatched(token, TokenType.StepLine, stepText, keyword, line.getIndent(), keywordType, null);
        return true;
    }

    @Override
    public boolean match_TableRow(Token token) {
        Line line = token.getRequiredLine();
        if (!line.startsWith(TABLE_CELL_SEPARATOR)) {
            return false;
        }
        List<LineSpan> tableCells = TableRowLine.parse(line.getIndent(), line.getText());
        setTokenMatched(token, TokenType.TableRow, null, null, line.getIndent(), null, tableCells);
        return true;
    }

    private String removeDocStringIndent(Token token) {
        Line line = token.getRequiredLine();
        if (activeDocStringSeparator == null) {
            return line.getRawText();
        }
        if (indentToRemove > line.getIndent()) {
            // BUG: Removes trailing whitespace!
            return line.getText();
        }
        return line.getRawTextSubstring(indentToRemove);
    }

    private String unescapeDocString(String text) {
        if (DOCSTRING_SEPARATOR.equals(activeDocStringSeparator)) {
            return text.replace("\\\"\\\"\\\"", DOCSTRING_SEPARATOR);
        }

        if (DOCSTRING_ALTERNATIVE_SEPARATOR.equals(activeDocStringSeparator)) {
            return text.replace("\\`\\`\\`", DOCSTRING_ALTERNATIVE_SEPARATOR);
        }
        return text;
    }
}

File: dart/lib/src/language/gherkin_token_matcher.dart
import 'package:cucumber_gherkin/src/exceptions/exceptions.dart';
import 'package:cucumber_gherkin/src/language/gherkin_language_constants.dart';
import 'package:cucumber_gherkin/src/language/gherkin_language_keywords.dart';
import 'package:cucumber_gherkin/src/language/gherkin_line.dart';
import 'package:cucumber_gherkin/src/language/location.dart';
import 'package:cucumber_gherkin/src/language/token.dart';
import 'package:cucumber_gherkin/src/parser/parser.g.dart';
import 'package:cucumber_gherkin/src/parser/token_matcher.dart';
import 'package:cucumber_messages/cucumber_messages.dart' as messages;

/// The [TokenMatcher] for plain (`.feature`) Gherkin sources.
class GherkinTokenMatcher implements TokenMatcher {
  /// Creates a matcher over the given dialects map, defaulting to English
  /// (`en`) unless another default language is supplied.
  GherkinTokenMatcher(this._dialects, [this._defaultLanguage = 'en']) {
    _useLanguage(
      _defaultLanguage,
      _lookupKeywords(_dialects, _defaultLanguage),
    );
  }
  static final RegExp _languagePattern = RegExp(
    r'^\s*#\s*language\s*:\s*([a-zA-Z\-_]+)\s*$',
  );

  final Map<String, GherkinLanguageKeywords> _dialects;
  final String _defaultLanguage;
  final Map<String, messages.StepKeywordType> _keywordTypesMap =
      <String, messages.StepKeywordType>{};

  late String _currentLanguage;
  late GherkinLanguageKeywords _currentKeywords;
  late List<String> _stepKeywordsByLengthDesc;
  String _activeDocStringSeparator = '';
  int _indentToRemove = 0;

  @override
  void reset() {
    _activeDocStringSeparator = '';
    _indentToRemove = 0;
    _useLanguage(
      _defaultLanguage,
      _lookupKeywords(_dialects, _defaultLanguage),
    );
  }

  @override
  bool matchEof(Token token) {
    if (token.isEof) {
      _setTokenMatched(token, TokenType.eof);
      return true;
    }
    return false;
  }

  @override
  bool matchOther(Token token) {
    // Take the whole line, minus DocString indents.
    final text = token.line.getLineText(_indentToRemove);
    _setTokenMatched(
      token,
      TokenType.other,
      text: _unescapeDocString(text),
      indent: 0,
    );
    return true;
  }

  @override
  bool matchEmpty(Token token) {
    if (token.line.isEmptyLine) {
      _setTokenMatched(token, TokenType.empty);
      return true;
    }
    return false;
  }

  @override
  bool matchComment(Token token) {
    if (token.line.startsWith(commentPrefix)) {
      final text = token.line.getLineText();
      _setTokenMatched(token, TokenType.comment, text: text, indent: 0);
      return true;
    }
    return false;
  }

  @override
  bool matchLanguage(Token token) {
    final match = _languagePattern.firstMatch(token.line.getLineText());
    if (match != null) {
      final language = match.group(1) ?? '';
      _setTokenMatched(token, TokenType.language, text: language);
      _useLanguage(
        language,
        _lookupKeywords(_dialects, language, token.location),
      );
      return true;
    }
    return false;
  }

  @override
  bool matchTagLine(Token token) {
    if (token.line.startsWith(tagPrefix)) {
      _setTokenMatched(token, TokenType.tagLine, items: token.line.tags);
      return true;
    }
    return false;
  }

  @override
  bool matchFeatureLine(Token token) =>
      _matchTitleLine(token, TokenType.featureLine, _currentKeywords.feature);

  @override
  bool matchRuleLine(Token token) =>
      _matchTitleLine(token, TokenType.ruleLine, _currentKeywords.rule);

  @override
  bool matchBackgroundLine(Token token) => _matchTitleLine(
    token,
    TokenType.backgroundLine,
    _currentKeywords.background,
  );

  @override
  bool matchScenarioLine(Token token) =>
      _matchTitleLine(
        token,
        TokenType.scenarioLine,
        _currentKeywords.scenario,
      ) ||
      _matchTitleLine(
        token,
        TokenType.scenarioLine,
        _currentKeywords.scenarioOutline,
      );

  @override
  bool matchExamplesLine(Token token) =>
      _matchTitleLine(token, TokenType.examplesLine, _currentKeywords.examples);

  bool _matchTitleLine(
    Token token,
    TokenType tokenType,
    List<String> keywords,
  ) {
    for (final keyword in keywords) {
      if (token.line.startsWithTitleKeyword(keyword)) {
        final title = token.line.getRestTrimmed(
          keyword.length + titleKeywordSeparator.length,
        );
        _setTokenMatched(token, tokenType, keyword: keyword, text: title);
        return true;
      }
    }
    return false;
  }

  @override
  bool matchDocStringSeparator(Token token) =>
      _activeDocStringSeparator.isEmpty
          // open
          ? _matchDocStringSeparator(token, docStringSeparator, true) ||
              _matchDocStringSeparator(
                token,
                docStringAlternativeSeparator,
                true,
              )
          // close
          : _matchDocStringSeparator(token, _activeDocStringSeparator, false);

  bool _matchDocStringSeparator(Token token, String separator, bool isOpen) {
    if (token.line.startsWith(separator)) {
      var mediaType = '';
      if (isOpen) {
        mediaType = token.line.getRestTrimmed(separator.length);
        _activeDocStringSeparator = separator;
        _indentToRemove = token.line.indent;
      } else {
        _activeDocStringSeparator = '';
        _indentToRemove = 0;
      }
      _setTokenMatched(
        token,
        TokenType.docStringSeparator,
        text: mediaType,
        keyword: separator,
      );
      return true;
    }
    return false;
  }

  @override
  bool matchStepLine(Token token) {
    for (final keyword in _stepKeywordsByLengthDesc) {
      if (token.line.startsWith(keyword)) {
        final stepText = token.line.getRestTrimmed(keyword.length);
        _setTokenMatched(
          token,
          TokenType.stepLine,
          keyword: keyword,
          text: stepText,
          keywordType:
              _keywordTypesMap[keyword] ?? messages.StepKeywordType.unknown,
        );
        return true;
      }
    }
    return false;
  }

  @override
  bool matchTableRow(Token token) {
    if (token.line.startsWith(tableCellSeparator)) {
      _setTokenMatched(token, TokenType.tableRow, items: token.line.tableCells);
      return true;
    }
    return false;
  }

  /// Records that [token] matched [matchedType], populating its matched
  /// keyword, text, items, indentation, and location.
  void _setTokenMatched(
    Token token,
    TokenType matchedType, {
    String text = '',
    String keyword = '',
    messages.StepKeywordType? keywordType,
    int? indent,
    Iterable<GherkinLineSpan> items = const <GherkinLineSpan>[],
  }) {
    final matchedIndent =
        matchedType == TokenType.empty ? 0 : indent ?? token.line.indent;
    token
      ..matchedType = matchedType
      ..matchedKeyword = keyword
      ..matchedKeywordType = keywordType
      ..matchedText = text
      ..matchedItems = items
      ..location = Location(token.location.line, matchedIndent + 1);
    // Only Feature lines carry the active dialect into the AST builder.
    if (matchedType == TokenType.featureLine) {
      token.matchedLanguage = _currentLanguage;
    }
  }

  void _useLanguage(String language, GherkinLanguageKeywords keywords) {
    _currentLanguage = language;
    _currentKeywords = keywords;
    // Longest first so a short keyword that is a prefix of a longer one cannot
    // steal the match.
    _stepKeywordsByLengthDesc =
        <String>{
            ...keywords.given,
            ...keywords.when,
            ...keywords.then,
            ...keywords.and,
            ...keywords.but,
          }.toList()
          ..sort((a, b) => b.length - a.length);
    _initializeKeywordTypes(keywords);
  }

  String _unescapeDocString(String text) {
    if (docStringSeparator == _activeDocStringSeparator) {
      return text.replaceFirst(r'\"\"\"', docStringSeparator);
    }
    if (docStringAlternativeSeparator == _activeDocStringSeparator) {
      return text.replaceFirst(r'\`\`\`', docStringAlternativeSeparator);
    }
    return text;
  }

  void _initializeKeywordTypes(GherkinLanguageKeywords keywords) {
    _keywordTypesMap.clear();
    void mapKeywords(
      Iterable<String> stepKeywords,
      messages.StepKeywordType type,
    ) {
      for (final keyword in stepKeywords) {
        final existing = _keywordTypesMap[keyword];
        // Ambiguous keywords (e.g. `* ` in given/when/then/and) are unknown.
        _keywordTypesMap[keyword] =
            existing == null || existing == type
                ? type
                : messages.StepKeywordType.unknown;
      }
    }

    mapKeywords(keywords.given, messages.StepKeywordType.context);
    mapKeywords(keywords.when, messages.StepKeywordType.action);
    mapKeywords(keywords.then, messages.StepKeywordType.outcome);
    mapKeywords(keywords.and, messages.StepKeywordType.conjunction);
    mapKeywords(keywords.but, messages.StepKeywordType.conjunction);
  }

  static GherkinLanguageKeywords _lookupKeywords(
    Map<String, GherkinLanguageKeywords> dialects,
    String language, [
    Location location = Location.empty,
  ]) {
    final keywords = dialects[language];
    if (keywords == null) {
      throw NoSuchLanguageException(language, location);
    }
    return keywords;
  }
}
